src/software/geom/util.cpp

```diff
@@ -1251,6 +1116,172 @@ std::optional<Segment> mergeFullyOverlappingSegments(Segment segment1, Segment s
+    {
+        if (!collinear(segments[i], segments[i + 1]))
+        {
+            return std::nullopt;
+        }
+    }
+    std::vector<Segment> unique_segments;
+
+    unsigned int j = 0;
+    // Loop through all segments and combine segments
+    // to reduce the vector to the smallest number of independent (not overlapping)
+    // segments
+    while (segments.size() > 0)
+    {
+        std::optional<Segment> temp_segment = segments[0];
+        unique_segments.push_back(temp_segment.value());
+        segments.erase(segments.begin());
+
+        for (unsigned int i = 0; i < segments.size(); i++)
+        {
+            temp_segment =
+                mergeOverlappingParallelSegments(unique_segments[j], segments[i]);
+
+            if (temp_segment.has_value())
+            {
+                unique_segments[j] = temp_segment.value();
+                // Remove segments[i] from the list as it is not unique
+                segments.erase(segments.begin() + i);
+                i--;
+            }
+        }
+        j++;
+    }
+
+    return unique_segments;
+}
+
+Shot calcMostOpenDirection(Point origin, Segment segment, std::vector<Circle> obstacles)
+{
+    std::vector<Segment> obstacle_segment_projections;
+    // If there are no obstacles, return the center of the Segment and the shot angle
+    if (obstacles.size() == 0)
+    {
+        return Shot(Point::getAverageOfPoints({segment.getSegStart(), segment.getEnd()}),
+                    ((segment.getSegStart() - origin)
+                         .orientation()
+                         .minDiff((segment.getEnd() - origin).orientation()))
+                        .abs());
+    }
+
+    // Loop through all obstacles to create their 'blocking' Segment
+    for (Circle circle : obstacles)
+    {
+        // If the reference is inside an obstacle there is no open direction
+        if (contains(circle, origin))
+        {
+            return Shot(Point(segment.getSegStart() +
+                              (segment.getEnd() - segment.getSegStart()) / 2),
+                        Angle::fromDegrees(0));
+        }
+
+        // Get the tangent rays from the reference point to the obstacle
+        auto [ray1, ray2] = getCircleTangentRaysToReference(origin, circle);
+
+        // Project the tangent Rays to obtain a 'blocked' segment on the reference Segment
+        std::optional<Segment> intersect_segment =
+            getIntersectingSegment(ray1, ray2, segment);
+
+        if (intersect_segment.has_value())
+        {
+            obstacle_segment_projections.push_back(intersect_segment.value());
+        }
+    }
+
+    // If we have more than 1 Segment from the obstacle projection then we must combine
+    // overlapping ones to simplify analysis
+    if (obstacle_segment_projections.size() >= 2)
+    {
+        obstacle_segment_projections =
+            reduceParallelSegments(obstacle_segment_projections).value();
+    }
```

Comment:
Wouldn't all the projection segments be parallel? So this would always return an empty vector?

Reply:
Ah yeah, reduceParallelSegments takes in a vector of Segments and returns a vector with the fewest, independent Segments that cover the same linear space.

The javadoc comment for the function was wrong. I fixed it and added some ascii art to explain